File: csharp/src/ConsoleForum.App/Commands/DemoteCommand.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using ConsoleForum.App.Commands.Attributes;
using ConsoleForum.App.Common;
using ConsoleForum.App.Core;
using ConsoleForum.App.Database;
using ConsoleForum.App.IO;
using ConsoleForum.App.Models;
using ConsoleForum.App.Models.Enums;

namespace ConsoleForum.App.Commands
{
    [AdminCommand]
    [CommandHelper("Demotes the User with the given Id into a lower role.")]
    [CommandExample("Demote|1")]
    public class DemoteCommand : ICommand
    {
        private const string NoSuchUserMessage = "There is no User with the given Id.";

        private const string SuccessDemoteMessage = "Successfully Demoted User - {0}";

        public IInputReader ConsoleReader { get; set; }

        public IOutputWriter ConsoleWriter { get; set; }

        public List<string> Parameters { get; set; }

        public IPrincipal Principal { get; set; }

        public bool ResetsConsole { get; set; } = true;

        public void Execute()
        {
            int userId = int.Parse(this.Parameters[0]);

            User userFromDb = DbContext.Users.FirstOrDefault(user => user.Id == userId);

            if (userFromDb == null || userFromDb.IsDeleted)
            {
                this.ConsoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                this.ConsoleWriter.ErrorLine(NoSuchUserMessage);
                Thread.Sleep(Constants.NotificationDelay);
            }
            else
            {
                if (userFromDb.Role == UserRole.Admin)
                {
                    userFromDb.Role = UserRole.Moderator;
                }
                else if (userFromDb.Role == UserRole.Moderator)
                {
                    userFromDb.Role = UserRole.User;
                }

                this.ConsoleWriter.Write(Constants.ConsoleForumOutputPrefix);
                this.ConsoleWriter.SuccessLine(SuccessDemoteMessage, userId);
                Thread.Sleep(Constants.NotificationDelay);
            }
        }
    }
}

File: java/src/org/sanity/consoleForum/commands/DemoteCommand.java
package org.sanity.consoleForum.commands;

import org.sanity.consoleForum.commands.Attributes.AdminCommand;
import org.sanity.consoleForum.commands.Attributes.CommandExample;
import org.sanity.consoleForum.commands.Attributes.CommandHelper;
import org.sanity.consoleForum.common.Constants;
import org.sanity.consoleForum.core.Principal;
import org.sanity.consoleForum.database.EntityManager;
import org.sanity.consoleForum.io.InputReader;
import org.sanity.consoleForum.io.OutputWriter;
import org.sanity.consoleForum.models.User;
import org.sanity.consoleForum.models.enums.UserRole;

import java.io.IOException;
import java.util.List;

@AdminCommand
@CommandHelper(help = "Demotes the User with the given Id into a lower role.")
@CommandExample(example = "Demote|1")
public class DemoteCommand implements Command {
    private static final String NoSuchUserMessage = "There is no User with the given Id.";

    private static final String SuccessDemoteMessage = "Successfully Demoted User - %s";

    private InputReader consoleReader;

    private OutputWriter consoleWriter;

    private List<String> parameters;

    private Principal principal;

    private boolean resetsConsole;

    private EntityManager entityManager;

    @Override
    public InputReader getConsoleReader() {
        return this.consoleReader;
    }

    @Override
    public void setConsoleReader(InputReader inputReader) {
        this.consoleReader = inputReader;
    }

    @Override
    public OutputWriter getConsoleWriter() {
        return this.consoleWriter;
    }

    @Override
    public void setConsoleWriter(OutputWriter outputWriter) {
        this.consoleWriter = outputWriter;
    }

    @Override
    public List<String> getParameters() {
        return this.parameters;
    }

    @Override
    public void setParameters(List<String> parameters) {
        this.parameters = parameters;
    }

    @Override
    public Principal getPrincipal() {
        return this.principal;
    }

    @Override
    public void setPrincipal(Principal principal) {
        this.principal = principal;
    }

    @Override
    public boolean getResetsConsole() {
        return this.resetsConsole;
    }

    @Override
    public void setResetsConsole(boolean resetsConsole) {
        this.resetsConsole = resetsConsole;
    }

    @Override
    public EntityManager getEntityManager() {
        return this.entityManager;
    }

    @Override
    public void setEntityManager(EntityManager entityManager) {
        this.entityManager = entityManager;
    }

    public void execute() throws IOException, InterruptedException {
        int userId = Integer.parseInt(this.getParameters().get(0));

        User userFromDb = this.getEntityManager().getUsers()
                .stream()
                .filter(user -> user.getId() == userId)
                .findFirst()
                .orElse(null);

        if (userFromDb == null || userFromDb.getIsDeleted())
        {
            this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
            this.consoleWriter.errorLine(NoSuchUserMessage);
            Thread.sleep(Constants.NOTIFICATION_DELAY);
        }
        else
        {
            if (userFromDb.getRole() == UserRole.ADMIN)
            {
                userFromDb.setRole(UserRole.MODERATOR);
            }
            else if (userFromDb.getRole() == UserRole.MODERATOR)
            {
                userFromDb.setRole(UserRole.USER);
            }

            this.consoleWriter.write(Constants.CONSOLE_FORUM_OUTPUT_PREFIX);
            this.consoleWriter.successLine(SuccessDemoteMessage, userId);
            Thread.sleep(Constants.NOTIFICATION_DELAY);
        }
    }
}
